app/src/main/java/pm/gnosis/heimdall/ui/safe/pairing/PairingAuthenticator.kt

```
@@ -0,0 +1,193 @@
+import timber.log.Timber
+import java.math.BigInteger
+import java.math.RoundingMode
+import javax.inject.Inject
+
+abstract class PairingAuthenticatorContract : ViewModel() {
+
+    abstract val observableState: LiveData<ViewUpdate>
+
+    abstract fun setup(safeAddress: Solidity.Address)
+
+    abstract fun loadAuthenticatorInfo()
+
+    abstract fun estimate()
+
+    /**
+     * @param signingSafe to indicate which key should be used for pairing. @null if a new key should be generated
+     */
+    abstract fun pair(payload: String)
+
+    sealed class PairingResult {
+
+        data class PairingError(val error: Exception) : PairingResult()
+
+        data class PairingSuccess(val setpInfo: AuthenticatorSetupInfo) : PairingResult()
+    }
+
+    sealed class ViewUpdate {
+
+        data class Balance(
+            val balanceBefore: String,
+            val fee: String,
+            val balanceAfter: String,
+            val tokenSymbol: String,
+            val sufficient: Boolean
+        ) : ViewUpdate()
+
+        data class Pairing(
+            val isLoading: Boolean,
+            val pairingResult: PairingResult? = null
+        ) : ViewUpdate()
+
+        data class Authenticator(
+            val info: AuthenticatorSetupInfo
+        ) : ViewUpdate()
+    }
+
+}
+
+class PairingAuthenticatorViewModel @Inject constructor(
+    private val recoverSafeOwnersHelper: RecoverSafeOwnersHelper,
+    private val gnosisSafeRepository: GnosisSafeRepository,
+    private val accountsRepository: AccountsRepository,
+    private val tokenRepository: TokenRepository,
+    private val transactionExecutionRepository: TransactionExecutionRepository,
+    private val pushServiceRepository: PushServiceRepository,
+    private val moshi: Moshi,
+    private val appDispatcher: ApplicationModule.AppCoroutineDispatchers
+) : PairingAuthenticatorContract() {
+
+    private val errorHandler = CoroutineExceptionHandler { _, e ->
+        viewModelScope.launch {
```

Should we propagate the error to the user?